posix + kernel: support for realtime signals

cfriedt opened the pull request:
This was required for practicality
* ring_buffer: support static rings

These add support to the kernel
* kernel: initial support for signals
* tests: kernel: add tests for `k_sigqueue()` and `k_sigtimedwait()`
* doc: kernel: services: add a doc about signals

These complete the `POSIX_REALTIME_SIGNALS` Option Group (and `_POSIX_REALTIME_SIGNALS` Option)
* posix: add realtime signal support
* tests: posix: common: add tests for realtime signals
* doc: posix: mark realtime signals as supported 🎉

These add `pthread_kill()` to (finally) complete the `POSIX_THREADS_BASE` Option Group and `_POSIX_THREADS` Option
* posix: threads: implement `pthread_kill()`
* tests: posix: headers: test for `pthread_kill()`
* doc: posix: mark posix threads as supported 🎉

[Kernel Doc Preview](https://builds.zephyrproject.io/zephyr/pr/74918/docs/kernel/services/signals.html)
[POSIX Doc Preview](https://builds.zephyrproject.io/zephyr/pr/74918/docs/services/portability/posix/index.html)

Fixes #59943
Fixes #74993
Fixes #74994
Fixes #74995

Commit:
doc: posix: mark realtime signals as supported

Mark the POSIX_REALTIME_SIGNALS Option Group (and Option) as
supported in Zephyr.

This and POSIX_THREAD_SAFE_FUNCTIONS (as well as the other
PRs for POSIX_DEVICE_IO, POSIX_FD_MGMT, and POSIX_SIGNALS)
should make us functionally complete for the required system
interfaces as well as PSE51.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: realtime signals: add realtime signal support

Add support for the _POSIX_REALTIME_SIGNALS Option.

This adds support for the functions sigqueue(),
sigtimedwait(), and sigwaitinfo().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
kernel: initial support for signals

This changes adds initial support for signals.

Note: this feature is experimental. The primary use-case
is intended to be real-time signals. Currently, there is
no asynchronous signal delivery.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
ring_buffer: support static rings

Move initializers to a common macro and provide
static variants.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: common: add tests for realtime signals

Add tests for realtime signals.

Note, that since the "is a" relationship between Zephyr and
POSIX threads is not bidirectional, tests need to run inside of
a POSIX thread in order to correctly translate a "pid_t"
(which is necessarily an int due to it being defined as such
in various C libraries) to a pthread_t, and subsequently to
a k_tid_t.

Of course, the kernel counterparts of the real-time signal API
have no such restriction, as k_pid_t aliases k_tid_t.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: kernel: add tests for k_sig_queue() and k_sig_timedwait()

Add tests for realtime signals.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Works on my desk ¯\_(ツ)_/¯
```shell
twister --retry-failed 3 -T tests/posix/common -T tests/kernel/signal/
..
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-rc1-12-gc284b50da617
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  562/ 562  100%  skipped:  272, failed:    0, error:    0
INFO    - 16 test scenarios (562 test instances) selected, 272 configurations skipped (246 by static filter, 26 at runtime).
INFO    - 290 of 562 test configurations passed (100.00%), 0 failed, 0 errored, 272 skipped with 0 warnings in 696.02 seconds
INFO    - In total 23167 test cases were executed, 27603 skipped on 41 out of total 757 platforms (5.42%)
INFO    - 288 test configurations executed on platforms, 2 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

Commit:
posix: threads: implement pthread_kill()

Now that POSIX_REALTIME_SIGNALS is implemented, add the final
piece of the POSIX_THREADS puzzle. Namely, pthread_kill().

This was the last function required to mark the
POSIX_THREADS_BASE Option Group as fully supported in Zephyr.

Additionally, we may now mark the _POSIX_THREADS Option as
supported in Zephyr. The _POSIX_THREADS Option is required
as part of the POSIX System Interfaces (Base Definitions)
as well as PSE51, PSE52, PSE53, PSE54 and all other conformant
POSIX subprofiles.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: posix: mark posix threads as supported

Mark the POSIX_THREADS_BASE Option Group as complete and
mark the _POSIX_THREADS Option as supported.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: headers: test for pthread_kill()

Test for pthread_kill() in signal.h

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

nashif commented:
@cfriedt  Quickly looked at this, some comments:

- there is a layer violation in the kernel, i.e. you have CONFIG_POSIX_SIGNALS inside kernel/ and dependency on it in the Kconfig.
- You are introducing a new API into kernel.h `k_sig_*` without any reference or documentation what this does and how it can be used standalone without posix portability layer being enabled.

cfriedt commented:
> @cfriedt Quickly looked at this, some comments:
> 
> * there is a layer violation in the kernel, i.e. you have CONFIG_POSIX_SIGNALS inside kernel/ and dependency on it in the Kconfig.

I can see why it would appear at first that there might be a layering violation. However, there is actually none; the kernel is the logical common place to collect signal support, since it is required for ISO C, POSIX, and Realtime signals. In this case, the kernel Kconfig integer is really just acting as an "accumulator" for a number of other Kconfig integers spread throughout other parts of Zephyr.

As far as I know, there are no other APIs that use the same signal concept, but if there were, they would also just be "accumulated" in the same fashion. I.e. it is consistent.

It might be more obvious after first glance, but in the actual C code introduced here, Zephyr does not depend on or reference anything from POSIX (it's the exclusively the opposite). So the code dependency graph is unidirectional (aside from the POSIX features that have already been baked into Zephyr, such as `errno`, `strnlen()`, etc), and therefore there is no layering violation.

> * You are introducing a new API into kernel.h `k_sig_*` without any reference or documentation what this does and how it can be used standalone without posix portability layer being enabled.

There is Doxygen and there are tests. It certainly does not require POSIX to be enabled, as evident from the tests. Additionally, it can be used by any ISO C library wishing to provide support for signals, but that is out of scope of the POSIX API and not part of this PR.

I suppose I could write a page about it, if that's necessary.

abrodkin commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+ * @retval 0 on success
+ * @retval -EINVAL for invalid values of @a how
+ */
+__syscall int k_sig_mask(int how, const struct k_sig_set *set, struct k_sig_set *oset);
```
`k_sigmask()`?

cfriedt replied:
Since Jukka also thought `k_sig_` was maybe preferable, I've left it that way (can revisit if needed).

ycsin replied:
+1 to `k_sig_*`

abrodkin commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+ * @retval 0 on success
+ * @retval -EINVAL for invalid values of @a how
+ */
+__syscall int k_sig_pending(struct k_sig_set *set);
```
`k_sigpending()`?

cfriedt replied:
Since Jukka also thought `k_sig_` was maybe preferable, I've left it that way (can revisit if needed).

abrodkin commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+ * @return -EPERM if the calling thread lacks permissions to signal the process with id @a pid.
+ * @return -ESRCH if a process with id @a pid does not exist.
+ */
+__syscall int k_sig_queue(k_pid_t pid, int signo, union k_sig_val value);
```
`k_sigqueue()`?

cfriedt replied:
Since Jukka also thought `k_sig_` was maybe preferable, I've left it that way (can revisit if needed).

abrodkin commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+ *
+ * @see CONFIG_DYNAMIC_THREAD
+ */
+__syscall int k_sig_timedwait(const struct k_sig_set *set, struct k_sig_info *info,
```
`k_sigtimedwait()`?

cfriedt replied:
Since Jukka also thought `k_sig_` was maybe preferable, I've left it that way (can revisit if needed).

abrodkin reviewed (commented):
Amazing as usual! Makes POSIX layer of Zephyr even more complete, thus simplifying porting of software - that's much appreciated!

Though I think it makes sense to be closer to the standard we're implementing here, thus public functions names and definitions IMHO worth aligning with the standard.

npitre commented:
Documentation for `k_sig_emptyset()`seems wrong (looks like a copy of
the one for `k_sig_fillset()`).


This looks odd:

```
struct k_sig_set {
       unsigned long sig[ROUND_UP(CONFIG_SIGNAL_SET_SIZE, BITS_PER_LONG) / BITS_PER_LONG];
};
```

I'd suggest `DIV_ROUND_UP(CONFIG_SIGNAL_SET_SIZE, BITS_PER_LONG)` instead.
Same thing with `K_STRSIGSET_CHARS`.


You have a mix of:

```
        for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) { ... }
```

and:

```
        ARRAY_FOR_EACH(set->sig, i) { ... }
```

Maybe uniformize those.

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+#define K_SIG_UNBLOCK 2
+
+/** @brief Minimum realtime signal number */
+#define K_SIG_RTMIN 32
```
There is actually a really amazing reason *not* to define `K_SIG_RTMAX` (or `K_SIRGTMAX`).

As mentioned in the POSIX spec, `[min, max]` is inclusive. Also mentioned elsewhere, is that developers are strongly encouraged to only ever use `min + i` when looping over RT sets.

That is because, max is defined in terms of min and the number of signals (`max = min + N`).

Intuitively, the way that one would normally go about writing an inclusive for loop is with `for (i= min; i <= max; i++)`. However, if there are zero RT signals, max is still defined as `min + N = min`. The loop is evaluated (a bug).

Now, if max is not defined, then the usual approach for writing a for loop would be `for (i = 0; i < N; i++) .. x = (min + i)..` or `for (i = min; i < min + N; i++)` .  if `N = 0`, the loop is not evaluated (no bug).

I actually just debugged all of this so that others wouldn't have to later 😅

abrodkin replied:
Makes sense now! Then maybe add that note in the code/comments so that some smart developer in the future won't even attempt "improving" that implementation?

cfriedt replied:
Fixed!

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+#define K_SIG_UNBLOCK 2
+
+/**
+ * @brief Add a signal @a sig to signal set @a set.
```
I have no hard preference, and could definitely switch to `k_sig` instead of `k_sig_`. Whatever is easier to remember.

Definitely also happy to improve docs.

jukkar replied:
If we want to namespace this API, then it might be better to stick with `k_sig_xxx`

cfriedt commented:
> Documentation for `k_sig_emptyset()`seems wrong (looks like a copy of
> the one for `k_sig_fillset()`).

I'll take a look.

> I'd suggest `DIV_ROUND_UP(CONFIG_SIGNAL_SET_SIZE, BITS_PER_LONG)` instead.
> Same thing with `K_STRSIGSET_CHARS`.

I tried that originally. On revisiting, it seems that neither of those work in the case the numerator is zero, so I will add a `MAX()` operation as well.

> ```
>         ARRAY_FOR_EACH(set->sig, i) { ... }
> ```
> 
> Maybe uniformize those.

Will do

nashif commented:
> > * there is a layer violation in the kernel, i.e. you have CONFIG_POSIX_SIGNALS inside kernel/ and dependency on it in the Kconfig.
> 
> I can see why it would appear at first that there might be a layering violation. However, there is actually none; the kernel is the logical common place to collect signal support, since it is required for ISO C, POSIX, and Realtime signals. In this case, the kernel Kconfig integer is really just acting as an "accumulator" for a number of other Kconfig integers spread throughout other parts of Zephyr.
> 
> As far as I know, there are no other APIs that use the same signal concept, but if there were, they would also just be "accumulated" in the same fashion. I.e. it is consistent.
> 
> It might be more obvious after first glance, but in the actual C code introduced here, Zephyr does not depend on or reference anything from POSIX (it's exclusively the opposite). So the code dependency graph is unidirectional (aside from the POSIX features that have already been baked into Zephyr, such as `errno`, `strnlen()`, etc), and therefore there is no layering violation.

I was referring to the Kconfig above, not the code. It would be nice to not have CONFIG_POSIX in the kernel/ to avoid any confusion, probably there is a better way to set the queue size other than referencing a kconfig from 1 layer up.

> There is Doxygen and there are tests. It certainly does not require POSIX to be enabled, as evident from the tests. Additionally, it can be used by any ISO C library wishing to provide support for signals, but that is out of scope of the POSIX API and not part of this PR.
> 
> I could write a page about it, if that's necessary, but it might be better to defer that given that it is experimental.

Docygen is not enough, standalone usage of signals needs to be detailed, similar to all other kernel premitives and services we currently have and the usecase (standalone) needs to be provided.

cfriedt commented:
> I was referring to the Kconfig above, not the code. It would be nice to not have CONFIG_POSIX in the kernel/ to avoid any confusion, probably there is a better way to set the queue size other than referencing a kconfig from 1 layer up.

Yeah, it's unfortunate that Kconfig does not allow `select` for integer Kconfig options, otherwise, it would be fairly straightforward.

Technically speaking, we could just have a default of 1 if `SIGNAL` is selected, otherwise 0, and then simply have build-time assertions in other areas saying like so:

```cpp
// signal.c (e.g. common libc)
BUILD_ASSERT(CONFIG_SIGNAL_SET_SIZE >= 6, "ISO C requires CONFIG_SIGNAL_SET_SIZE >= 6");
// signal.c (posix)
BUILD_ASSERT(CONFIG_SIGNAL_SET_SIZE >= 31, "POSIX signals require CONFIG_SIGNAL_SET_SIZE >= 31");
// rtsig.c (realtime)
BUILD_ASSERT(CONFIG_SIGNAL_SET_SIZE >= 32, "POSIX signals require CONFIG_SIGNAL_SET_SIZE >= 32 (40 to be conformant)");
```

> Docygen is not enough, standalone usage of signals needs to be detailed, similar to all other kernel premitives and services we currently have and the usecase (standalone) needs to be provided.

Ok. I'll add a sample and a page in the documentation.

ceolin commented on kernel/signal.c:
```diff
@@ -0,0 +1,552 @@
+static bool k_sig_setisempty(const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(set->sig, i) {
+		if (POPCOUNT(set->sig[i]) > 0) {
```
Can't you just check `if (set->sig[i] > 0)`

cfriedt replied:
Yes - fixed!

ceolin commented on kernel/signal.c:
```diff
@@ -0,0 +1,552 @@
+
+static const char *k_strsigset_r(const struct k_sig_set *set, char *buf, size_t len)
+{
+	if ((set == NULL) || (buf == NULL) || (len < (K_STRSIGSET_CHARS + 1))) {
```
As far as I could see the only place that this function is called ensures that `buf` is not `NULL` and `len` is correct.

cfriedt replied:
Fixed! (removed the function in favour of `LOG_HEXDUMP_DBG()`)

ceolin commented on kernel/signal.c:
```diff
@@ -0,0 +1,552 @@
+		if (N == 0) {
+			LOG_DBG("sigq: (empty)");
+		} else {
+			static uint8_t buf[CONFIG_SIGNAL_QUEUE_SIZE];
```
This makes this function not re-entrant. Is that ok ?

cfriedt replied:
Fixed (dropped the function for `LOG_HEXDUMP_DBG()`)

ceolin commented on lib/posix/options/rtsig.c:
```diff
@@ -0,0 +1,64 @@
+
+	if (timeout == NULL) {
+		to = K_FOREVER;
+	} else if (timeout->tv_sec == 0 && timeout->tv_nsec == 0) {
```
Do you really need this condition ? Isn't `K_NO_WAIT == K_MSEC(0)`  ?

cfriedt replied:
Yeah, this could be removed.

cfriedt replied:
Fixed!

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
+	int si_signo;
+	int si_code;
+	union k_sig_val si_value;
+};
```
Yeah, that might make better sense. There was a compile error if `k_sig_set` wasn't in `kernel_structs.h` for some reason.

cfriedt replied:
Fixed!

cfriedt replied:
Fixed!

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,552 @@
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
```
I have a change queued up with that already. Will update in a bit

cfriedt replied:
Fixed!

cfriedt commented:
rebased

Commit:
doc: kernel: services: add a doc about signals

Add a document describing the signal API.

Provide details on

* manipulating signal sets
* manipulating the signal mask of the current thread
* sending (queuing) a signal
* receiving (waiting-on) a signal

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,552 @@
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
```
Fixed!

cfriedt commented:
* used `ARRAY_FOR_EACH()` where possible
* removed `k_strsignal*()` in favour of `LOG_HEXDUMP_DBG()`
* remove use of `POPCOUNT()`
* moved structs and macros to `include/zephyr/kernel/signal.h`
* used `MAX(x, 1)` with `DIV_ROUND_UP()` so that the numerator would be non-zero
* added a page with documentation to `doc/kernel/services/signal.rst`

ycsin commented on doc/kernel/services/signals.rst:
```diff
@@ -0,0 +1,204 @@
+    merges any signals (real-time or otherwise).
+
+To address the latter point, real-time signals have a priority with respect to
+other real-time signals. However, the POSIX specification does not clarify the
```
is this supposed to be:
```suggestion
other standard signals. However, the POSIX specification does not clarify the
```

cfriedt replied:
No, POSIX only specifies priority between real-time signals. There is no specification that standard signals have higher or lower priority than real-time signals.

I think it's better to be unambiguous though and clearly make the distinction that real-time signals have priority over standard signals.

cfriedt commented:
* touched-up `signals.rst` / doxygen

ycsin commented on doc/kernel/services/signals.rst:
```diff
@@ -0,0 +1,209 @@
+Receiving Signals
+=================
+
+A thread my query for the signals pending delivery via
```
```suggestion
A thread may query for the signals pending delivery via
```

ycsin reviewed (commented):
LGTM, just some minor compliance errors

cfriedt commented on lib/posix/options/rtsig.c:
```diff
@@ -0,0 +1,62 @@
+	if (timeout == NULL) {
+		to = K_FOREVER;
+	} else {
+		to = K_MSEC(timeout->tv_sec * MSEC_PER_SEC + timeout->tv_nsec / NSEC_PER_MSEC);
```
Hmm... preferably not, since it's a non-standard function. Ideally it would be removed.

cfriedt commented:
* got rid of Kconfig compliance issues (odd that checkpatch.pl did not catch those, and no warnings were generated at all... were they false positives?)

cfriedt commented:
Yeah, it would appear that the Kconfig warnings are a false positive.

jukkar commented on include/zephyr/arch/common/ffs.h:
```diff
@@ -37,6 +37,30 @@ static ALWAYS_INLINE unsigned int find_msb_set(uint32_t op)
+
+static ALWAYS_INLINE unsigned int find_msb_set_64(uint64_t op)
+{
+	if (op == 0) {
```
There already exists `u64_count_leading_zeros` and `u64_count_trailing_zeros` (and 32-bit variants for these). Why do we need these functions, or can we call those functions here in order to avoid duplicate implementations?

cfriedt replied:
Good point - I think we likely don't even need `find_msb_set_32()`.

cfriedt replied:
Fixed!

cfriedt commented:
* used `u*_count_leading_zeros()` instead of `find_lsb_set_*()`
* used bitwise operators instead of expensive division & modulo in signal arithmetic
* undid the unnecessary change of `SIGUSR1` to `SIGURG`

cfriedt commented:
* fixed commit message typo: `k_sigqueue()` -> `k_sig_queue()`

cfriedt commented:
* removed reference to non-existent `timeout` parameter in `k_sig_queue()` doxygen

cfriedt commented:
* Fixed up references in `signals.rst`

cfriedt commented:
I'm still seeing weirdness in the Doxygen output. Not entirely sure how to fix this
<img width="248" alt="Screenshot 2024-08-10 at 8 39 18 AM" src="https://github.com/user-attachments/assets/2688984a-d064-49ba-acde-eef5d3e72421">

kartben commented:
> I'm still seeing weirdness in the Doxygen output. Not entirely sure how to fix this. @kartben - have you ever seen this before? 
> <img width="248" alt="Screenshot 2024-08-10 at 8 39 18 AM" src="https://github.com/user-attachments/assets/2688984a-d064-49ba-acde-eef5d3e72421">
> 
> 

Is there a reason you're using `@headerfile`? Looks like the likely culprit (not sure why tho)

cfriedt commented:
> Is there a reason you're using `@headerfile`? Looks like the likely culprit (not sure why tho)

Actually, I added `@headerfile` thinking that would solve the problem, but it didn't help. With or without it was the same.

Then I reversed where `@addtogroup` and `@defgroup` were for the `k_sig_apis` group, and it half-worked. It now shows `#include <signal.h>` which is slightly correct. I think it would be preferable to use `#include <zephyr/kernel.h>` since kernel headers shouldn't be included directly.

Ahh... I see it's the same situation already for spinlocks.

<img width="197" alt="Screenshot 2024-08-10 at 9 06 16 AM" src="https://github.com/user-attachments/assets/fead5670-7380-4154-b70b-55a3bf92fbe8">

At least it's not only a problem here, so I think my next push should be OK.

Have a nice weekend. Please take a look when the signals are green again, but no rush.

kartben commented:
@cfriedt *if* you want to use `@headerfile` tags (which I don't think you need, but I might be missing something), then double check the correct syntax for sharp brackets: https://www.doxygen.nl/manual/commands.html#cmdheaderfile

cfriedt commented:
@jukkar, @ycsin - please revisit when you have a moment.
@kartben - `@headerfile` isn't used anywhere else in Zephyr currently, so I would prefer to not introduce it. The `#include ...` bit seems to be generated by [SHOW_HEADERFILE](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/zephyr.doxyfile.in#L637)

ycsin reviewed (approved):
I have mixed feelings over the intro/use of `k_pid_t` in the Zephyr APIs since the kernel actually do not support processes and not sure if it ever will, so instead of having this 'catch' in the Zephyr kernel layer, it might also make sense for us to simply treat thread as thread in Zephyr APIs and simply address/mention this limitation in the POSIX layer, but otherwise LGTM

cfriedt commented:
@peter-mitsis , @andyross - would you be able to review this from the kernel side?

peter-mitsis reviewed (commented):
Years ago, Zephyr had a signal-like feature called _alerts_; however, this was removed in 2018 as no one was using it. This leads me to wonder, would it be better to have this entirely defined within the posix subsystem/library and not in the kernel?

nashif commented:
> Years ago, Zephyr had a signal-like feature called _alerts_; however, this was removed in 2018 as no one was using it. This leads me to wonder, would it be better to have this entirely defined within the posix subsystem/library and not in the kernel?

for reference: https://docs.zephyrproject.org/1.13.0/kernel/synchronization/alerts.html?highlight=alerts

cfriedt commented:
> would it be better to have this entirely defined within the posix subsystem/library and not in the kernel?

The primary reason is that, eventually, the same signal mechanism can be used by both libc and POSIX to signal threads (so that libc signals can be independent of POSIX).

The second reason is that (currently) Zephyr threads are not POSIX threads, so if it were to be only available at the POSIX thread level, then there would be an asymmetric way to signal. This simplifies and future-proofs signalling so it is bidirectional.

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,545 @@
+#include <zephyr/sys/ring_buffer.h>
+
+#ifndef BITS_PER_BYTE
+#define BITS_PER_BYTE 8
```
Yeah, same. I could factor that out into a separate commit, if that works.

cfriedt replied:
Made how https://github.com/zephyrproject-rtos/zephyr/pull/79479

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,545 @@
+		return false;
+	}
+
+	if (_current == tid) {
```
Thanks for the insight - what is the best way to get the thread associated with a given syscall? `k_current_get()`?

peter-mitsis replied:
If I am following the intent of the logic correctly, all we really need for this _fast path_ check is

```
if ((!arch_is_in_isr() && (_current == tid)) {
    return true;
}
```

Thus, if the operation was called by a thread, it is safe to test __current_ against tid. If it was invoked by an ISR, we fall through to the next check just as if the routine was called by a thread other than itself.

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,545 @@
+	if ((tid->base.thread_state & valid_states) != 0) {
+		return true;
+	}
+
```
That works - I'll make those changes.

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,545 @@
+
+static void k_sig_pending_unlocked(struct k_sig_set *set)
+{
+	struct k_thread *tid = _current;
```
What is the best call to make to determine the thread ID associated with a syscall while the syscall is executing? `k_current_get()`?

peter-mitsis replied:
When it comes to getting the current thread ID, __current_ can generally be used by kernel internals--I usually think of k_current_get() as being more application oriented.

I may have misread something, but it does look to me like this routine is ISR callable. It can be called by z_impl_k_sig_queue(), which in turn has documentation that explicitly states that it is safe to call in an ISR context. Given how it gets used, I suspect that all we would need is to make the k_sig_addset() also conditional upon the context not being an ISR.

```
if ((entry->tid == tid) && !arch_is_in_isr()) {
    k_sig_addset(set, entry->signo);
}
```

cfriedt replied:
It might be possible to call this with tid as a parameter, which, I think would solidify the use.

It is static here, so use of this function should be from behind a syscall, and not an ISR.

cfriedt commented on kernel/signal.c:
```diff
@@ -0,0 +1,545 @@
+	}
+}
+
+static inline void k_sig_dump_set(const struct k_sig_set *set, const char *label)
```
Yes - thanks. I'll fix that.

peter-mitsis commented:
@cfriedt - Not sure if there is anything you need from me at the moment on this PR, but I thought I would ping you just in case I have overlooked something.

cfriedt commented:
@peter-mitsis - I have to respin. Just in Vienna at LPC this week. I'll make sure to check about ISR vs syscall context.

cfriedt commented:
I'd like to get #79479 in before this (to factor out the `BITS_PER_BYTE` macro). I have respun with changes requested, but will push after 79479 is in.
